src/geode/geosciences/lso_input.cpp

```diff
@@ -0,0 +1,782 @@
+                    unique_id = nb_unique_vertices;
+                    point = read_point();
+                    vertex_mapping_.emplace_back();
+                    nb_unique_vertices++;
+                }
+                const auto id = solid_builder_->create_point( point );
+                vertex_id_->set_value( id, unique_id );
+                vertex_mapping_[unique_id].push_back( id );
+            } while( std::getline( file_, line_ )
+                     && absl::StrContains( line_, "VRTX" ) );
+            builder_.create_unique_vertices( nb_unique_vertices );
+        }
+
+        std::tuple< geode::Point3D, geode::index_t > read_shared_point() const
+        {
+            const auto tokens = get_tokens();
+            geode::index_t value;
+            auto ok = absl::SimpleAtoi( tokens[2], &value );
+            OPENGEODE_EXCEPTION( ok, "[LSOInput] Error while "
+                                     "reading shared point index" );
+            const auto unique_id = value - OFFSET_START;
+            return std::make_tuple( solid_->point( unique_id ), unique_id );
+        }
+
+        geode::Point3D read_point() const
+        {
+            geode::Point3D point;
+            const auto tokens = get_tokens();
+            for( const auto i : geode::Range{ 3 } )
+            {
+                double value;
+                auto ok = absl::SimpleAtod( tokens[i + 2], &value );
+                OPENGEODE_EXCEPTION( ok, "[LSOInput] Error while "
+                                         "reading point coordinates" );
+                point.set_value( i, value );
+            }
+            point.set_value( 2, crs_.z_sign * point.value( 2 ) );
+            return point;
+        }
+
+        void read_tetra()
+        {
+            do
+            {
+                const auto tokens = get_tokens();
+                std::array< geode::index_t, 4 > vertices;
+                for( const auto i : geode::Range{ 4 } )
+                {
+                    auto ok = absl::SimpleAtoi( tokens[i + 1], &vertices[i] );
+                    OPENGEODE_EXCEPTION(
+                        ok, "[LSOInput] Error while reading tetra" );
+                    vertices[i] -= OFFSET_START;
+                }
+                solid_builder_->create_tetrahedron( vertices );
+                std::getline( file_, line_ );
+            } while( std::getline( file_, line_ )
+                     && geode::detail::string_starts_with( line_, "TETRA" ) );
+            solid_builder_->compute_polyhedron_adjacencies();
+        }
+
+        void read_surfaces()
+        {
+            if( !geode::detail::string_starts_with( line_, "MODEL" ) )
+            {
+                geode::detail::goto_keyword( file_, "MODEL" );
+            }
+            facet_id_ = solid_->facet_attribute_manager()
+                            .find_or_create_attribute< geode::VariableAttribute,
+                                geode::uuid >( "facet_id", default_id_ );
```

Comment:
in the constructor?

Reply:
this one is needed only after the solid is complete. My goal to create it only here is to avoid unecessary resize of the attribute.

Reply:
Ok for this one